### DISCUSSION TITLE
Insanely annoying "persisted failure reverts" in Foundry test

### DISCUSSION THREAD

**Discussion by CH-coding-fire** [Cyfrin/foundry-full-course-cu#1959] [General]
The behavior is, when "forge test --match-test invariant_mustHaveMoreValueThanTotalSupply123", then it tells me `[FAIL. Reason: revert: ERC20: insufficient allowance]`  This is **OK**

If I do the same command as above, it tells me: `[FAIL. Reason: invariant_mustHaveMoreValueThanTotalSupply123 persisted failure revert]` This is **CONFUSING as FK**

I am stuck on this for **HOURS**. Because I was not aware of the above, that is the repeated failure will cause "persisted failure revert", I googled and ChatGPT but found **NOTHING**, why is that????? How should a good programmer approach this problem? Also from common sense, why don't it just tell me again `[FAIL. Reason: revert: ERC20: insufficient allowance]`, like I think most programming tools do so there is consistency

![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/67637342/84ef29ac-9bff-4f03-9cb0-f6e6e4b540af)

![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/67637342/8406f3e7-1dd0-4366-aecb-2cd5d0bc0faa)

![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/67637342/a021886d-89e6-4b18-abeb-e2462a8a5e7a)

**Comment by EngrPips** [comment]
Hello @CH-coding-fire, Please show your invariant and handler file.

**Reply by CH-coding-fire** [reply]
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/67637342/34628b39-1eb6-4252-b072-8fbbb413670a)
<img width="820" alt="image" src="https://github.com/Cyfrin/foundry-full-course-cu/assets/67637342/f756bed2-c4f1-4fb7-a716-decb90ab08a7">

**Reply by CH-coding-fire** [reply]
But I guess this is fine, as I already figure out Foundry's behavior, I just find this frustrating and thinking how could I figure out this faster next time

**Reply by EngrPips** [reply]
Do you mean you resolved the issue already? And it will be better to copy the code and send it as text instead of an image next time. Wrap the code in three backticks to give it a nice formatting.

**Reply by smilingkylan** [reply]
@CH-coding-fire How did you fix it? I am having the same issue

**Reply by EngrPips** [reply]
Hello @kylanhurt, You figured it out already, right?

**Reply by Dennnnny** [reply]
hey guys, I met exactly the same error message: `persisted failure revert` 
so ... is there any way to solve ???????

**Reply by EngrPips** [reply]
@Dennnnny, run `forge clean,` then re-run your test process. Let's see if that fix it.

**Reply by Dennnnny** [reply]
@EngrPips thanks, it works now !

**Reply by EngrPips** [reply]
Awesome, Keep crushing it.

**Reply by mustby** [reply]
Thanks guys! This saved me a lot of time...

**Reply by EngrPips** [reply]
Anytime buddy.

**Reply by yawarasuuna** [reply]
ty. `forge clean` saving the day again, guess I'll have to remember to run it every time after an error lol

**Reply by EngrPips** [reply]
> ty. `forge clean` saving the day again. Guess I'll have to remember to run it every time after an error, lol

Foundry sometimes plays pranks on us, lol.

**Comment by nzes1** [comment]
Happy 2025 @CH-coding-fire  ,

This issue stems from Foundry's **Replaying Failures** feature. According to the [Foundry Book](https://book.getfoundry.sh/forge/replay-testing), this feature is designed to incrementally replay the last test run failures by persisting them to disk. The intention is to ensure any previously encountered failures are prioritized in subsequent test runs, helping identify and fix regressions more efficiently. 

However, this behavior can sometimes feel "annoying" during development, especially if you're not expecting it. 

Here's a breakdown of how it works:

#### **How Replaying Failures Works**
1. **Fuzz Tests**  
   - All counterexamples (failures) are saved to `~/.foundry/cache/fuzz/failures`.
   - On subsequent runs, these failures are replayed before new test campaigns are started.
   - Entries in this file are **appended** with new failures from subsequent test runs.

2. **Invariant Tests**  
   - Failures from invariant tests are saved in individual files under:  
     `~/.foundry/cache/invariant/failures/{TEST_SUITE_NAME}/{INVARIANT_NAME}`.
   - Unlike fuzz tests, these files are replaced when a new counterexample is found.

#### **How to Avoid Persisted Failures**
Currently, the only way to avoid replaying failures is to manually run the following command before running your test(s):
```bash
forge clean
```
This clears the cache and prevents Foundry from replaying previously recorded failures.

#### **Possible Improvements**
It might be helpful for the Foundry team to introduce a configuration option in `foundry.toml` to enable or disable this feature dynamically. This would make it easier for users who wish to skip the replaying of failures without manually running `forge clean`.

#### **Final Thoughts**
While this feature is invaluable for catching regressions, it can indeed feel inconvenient in certain workflows. A toggle or more control over this behavior in `foundry.toml` would likely improve the developer experience significantly.

**Reply by EngrPips** [reply]
Phenomenal heads-up.

**Reply by LuftieTheAnonymous** [reply]
Thank you for your comment. I don't know if you've written that on your own or is it AI, but it's great explanation.

**Comment by jorgezerpa** [comment]
In my case I just added a new command to my makefile so it runs `forge clean` before run my invariant tests. But YES, it is extremely annoying  :( I expended a lot of time debugging this, then notice it was failing by the same error when run the test with -vvvv and checkout the trace
